lib/connection/connection.js

```diff
@@ -88,14 +89,38 @@ function Connection(context)
     });
   };
 
+
+  /**
+  * Callback for connect() used to establish a connection.
+  *
+  * @param {self} this object
+  * @param {Function} callback
+  *
+  * @returns {function}
+  */
+  function connectCallback(self, callback)
+  {
+    return function (err)
+    {
+      if (Parameters.getValue(Parameters.names.CLIENT_SESSION_KEEP_ALIVE))
+      {
+        self.keepalive = setInterval(self.heartbeat, Parameters.getValue(Parameters.names.CLIENT_SESSION_KEEP_ALIVE_HEARTBEAT_FREQUENCY) * 1000, self);
+      }
+      if (Util.isFunction(callback))
+      {
+        callback(Errors.externalize(err), self);
+      }
+    };
+  }
+
   /**
    * Establishes a connection if we aren't in a fatal state.
    *
    * @param {Function} callback
    *
    * @returns {Object} the connection object.
    */
-  this.connect = function (callback)
+  this.connect = async function (callback)
```

Comment:
Is making this async safe?

Reply:
This could be why connections are timing out

Reply:
Does having two connect() functions sound reasonable? An async connect() for external browser, and a non-async connect() for default, key-pair, and oauth

Reply:
Why do you need an async connect for external browser?

Reply:
External browser's authentication returns a promise since retrieving the SAML token from the browser is done asynchronously.

With the async connect, it will wait for external browser's authenticate and services.sf.connect to finish execution before returning the connection. If it's not async, the connection may be returned without authenticate and services.sf.connect finishing their execution and making the changes it does to the connection

Reply:
Passing services.sf.connect to authenticate() as a function parameter does remove the need for async on connect.
But executing a query afterwards fails, because a connection hasn't been established yet when the query is executed.

Reply:
Okay that sounds reasonable. Btw, can you also add test cases for connecting with an external browser using this driver?

Reply:
connect() will behave as before. And the async connect() is for external browser.
Sure, will add test cases for external browser connection